File: n_puzzle/get_neighbours.py
from typing import List
from n_puzzle.puzzle import Puzzle


def _cook_new_map(parent: Puzzle, zero_id: int, new_id: int) -> Puzzle:

    new_map = parent.board.copy()
    new_map[zero_id], new_map[new_id] = new_map[new_id], new_map[zero_id]

    return Puzzle(new_map, g=parent.g + 1, h=None, parent=parent)


def get_neighbours(parent: Puzzle) -> List[Puzzle]:
    """
    top - left - right - bottom
    """
    result = []
    zero_id = parent.board.index(0)

    if zero_id - parent.side_len >= 0:
        new_map = _cook_new_map(parent, zero_id, zero_id - parent.side_len)
        new_map.set_move('u')
        result.append(new_map)
    if zero_id % parent.side_len - 1 >= 0:
        new_map = _cook_new_map(parent, zero_id, zero_id - 1)
        new_map.set_move('l')
        result.append(new_map)
    if zero_id % parent.side_len + 1 < parent.side_len:
        new_map = _cook_new_map(parent, zero_id, zero_id + 1)
        new_map.set_move('r')
        result.append(new_map)
    if zero_id + parent.side_len < parent.puzzle_len:
        new_map = _cook_new_map(parent, zero_id, zero_id + parent.side_len)
        new_map.set_move('d')
        result.append(new_map)

    return result

File: cpp/get_neighbours.cpp
#include "n_puzzle.h"

static Puzzle*
cook_new_map(Puzzle *current, int empty_id, int new_id)
{
	Puzzle *out = new Puzzle(current->map);
	std::swap(out->map[empty_id], out->map[new_id]);
	out->g = current->g + 1;
	out->parent = current;
	return (out);
}

static int	locate_empty_tile(Puzzle *current)
{
	for (int i = 0; i < current->map.size(); ++i)
		if (current->map[i] == current->map.size() - 1)
			return (i);
	throw;
}

std::vector<Puzzle*>*
get_neighbours(Puzzle *current)
{
	std::vector<Puzzle*> *result = new std::vector<Puzzle*>();
	int empty_id = locate_empty_tile(current);

	if (empty_id - Puzzle::side_len >= 0)
		result->push_back(cook_new_map(current, empty_id, empty_id - Puzzle::side_len));
	if (empty_id % Puzzle::side_len - 1 >= 0)
		result->push_back(cook_new_map(current, empty_id, empty_id - 1));
	if (empty_id % Puzzle::side_len + 1 < Puzzle::side_len)
		result->push_back(cook_new_map(current, empty_id, empty_id + 1));
	if (empty_id + Puzzle::side_len < current->map.size())
		result->push_back(cook_new_map(current, empty_id, empty_id + Puzzle::side_len));

	return result;
}
